network/src/protocols/discovery/mod.rs

```diff
@@ -1,426 +1,364 @@
+                                .into_iter()
+                                .flat_map(|node| node.addresses.into_iter())
+                                .collect::<Vec<_>>();
+
+                            state.addr_known.extend(addrs.iter());
+                            state.received_nodes = true;
+
+                            self.addr_mgr.add_new_addrs(session.id, addrs);
+                        }
+                    }
+                }
+            }
+            None => {
+                if self
+                    .addr_mgr
+                    .misbehave(session.id, Misbehavior::InvalidData)
+                    .is_disconnect()
+                    && context.disconnect(session.id).is_err()
+                {
+                    debug!("disconnect {:?} send fail", session.id)
+                }
+            }
+        }
+    }
+
+    fn notify(&mut self, context: &mut ProtocolContext, _token: u64) {
+        let now = Instant::now();
+
+        let dynamic_query_cycle = self.dynamic_query_cycle.unwrap_or(ANNOUNCE_INTERVAL);
+        let addr_mgr = &self.addr_mgr;
+
+        // get announce list
+        let announce_list: Vec<_> = self
+            .sessions
+            .iter_mut()
+            .filter_map(|(id, state)| {
+                // send all announce addr to remote
+                state.send_messages(context, *id);
+                // check timer
+                state.check_timer(now, dynamic_query_cycle);
+
+                if state.announce {
+                    state.announce = false;
+                    state.last_announce = Some(now);
+                    if let RemoteAddress::Listen(addr) = &state.remote_addr {
+                        if addr_mgr.is_valid_addr(addr) {
+                            Some(addr.clone())
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
                 } else {
-                    warn!("other channel error: {:?}", err);
+                    None
+                }
+            })
+            .collect();
+
+        if !announce_list.is_empty() {
+            let mut rng = rand::thread_rng();
+            let mut remain_keys = self.sessions.keys().cloned().collect::<Vec<_>>();
+            for announce_multiaddr in announce_list {
+                remain_keys.shuffle(&mut rng);
+                for i in 0..2 {
+                    if let Some(key) = remain_keys.get(i) {
+                        if let Some(value) = self.sessions.get_mut(key) {
+                            trace!(
+                                ">> send {} to: {:?}, contains: {}",
+                                announce_multiaddr,
+                                value.remote_addr,
+                                value.addr_known.contains(&announce_multiaddr)
+                            );
+                            if value.announce_multiaddrs.len() < 10
```

Comment:
Hardcoded, use `ANNOUNCE_THRESHOLD` instead

```suggestion
                            if value.announce_multiaddrs.len() < ANNOUNCE_THRESHOLD
```